dev/Repeater/APITests/ViewManagerTests.cs

```diff
@@ -759,6 +759,76 @@ public void ValidateFocusMoveOnElementClearedWithUniqueIds()
+        // Thus, the DataContext we see in GetElementCore is the OLD datacontext.
+        // If that data context is not null, that means it did not get cleared when the element was recycled, which is the wrong behavior.
+        // To check if the clearing is working correctly, we log all data contexts, and check if all of them are null.
+        public void ValidateElementClearingClearsDataContext()
+        {
+            ItemsRepeater repeater = null;
+            MockRecyclingTestFactory elementFactory = null;
+            int elementClearingRaisedCount = 0;
+            ItemsRepeaterScrollHost scroller = null;
+            Log.Comment("Initialize ItemsRepeater");
+            RunOnUIThread.Execute(() =>
+            {
+                elementFactory = new MockRecyclingTestFactory();
+                elementFactory.RecyclePool = new RecyclePool();
+                elementFactory.Templates["Item"] = (DataTemplate)XamlReader.Load(
+                    @"<DataTemplate xmlns='http://schemas.microsoft.com/winfx/2006/xaml/presentation'> " + "<Button Content=\"{Binding}\" />" + @"</DataTemplate>");
+
+                repeater = CreateRepeater(Enumerable.Range(0, 100),
+                    elementFactory);
+
+                repeater.Layout = new StackLayout();
+
+                var scrollViewer = new ScrollViewer() { Content = repeater };
+                scroller = new ItemsRepeaterScrollHost() 
+                {
+                    Height = 200,
+                    Width = 200,
+                    ScrollViewer = scrollViewer
+                };
+
+                Content = scroller;
+                scroller.UpdateLayout();
+                repeater.ElementClearing += Repeater_ElementClearing;
+            });
+
+            IdleSynchronizer.Wait();
+
+            Log.Comment("Scrolling ItemsRepeater around to clear and recycle elements");
+            RunOnUIThread.Execute(() =>
+            {
+                scroller.ScrollViewer.ChangeView(null, 400, null, true);
+                scroller.ScrollViewer.ChangeView(null, 600, null, true);
+                scroller.ScrollViewer.ChangeView(null, 800, null, true);
+            });
+
+            IdleSynchronizer.Wait();
+
+            Log.Comment("Verify ItemsRepeater cleared data contexts correctly");
+            Verify.IsTrue(elementClearingRaisedCount > 0, "ItemsRepeater should have cleared some elements");
+            foreach (object o in elementFactory.DataContextsInGetElement)
+            {
+                // TL;DR; on why we check this:
+                // All DataContexts that are present on our returned ItemTemplate(Shim) get logged inside DataContextsInGetElement.
+                // If one of them is not null, that means the DataContext did not get cleared correctly which is a bug.
+                Verify.IsNull(o, "Datacontext should be null on a recycled ItemTemplate");
+            }
```

Comment:
Don't need these if we add this verification in MockElementFactory.ClearElementFunc delegate above.

Reply:
ClearElement gets called before we are clearing the DataContext of the element, that's why we can't verify it in ClearElement. 

Or should the order of operations change, i.e:
1. Raise ElementClearing
2. Clear DataContext
3. Ask ElementFactory to clear element

Reply:
If we change the order to the above, testing will look quite differently, however with the current order of events:
1. Raise ElementClearing
2. Ask ElementFactory to clear element (DataContext is still known)
3. Clear DataContext
[...]
4. Recycle Template <- If DataContext is still set here, that's a problem

We need to check if a recycled elements DataContext is null before we pass it to the ItemsRepeater.

Reply:
I mean MockElementFactory.ClearElementFunc delegate method, not ItemsRepeater.ClearElement event handler. When the MockElementFactory gets the element after being recycled it should be null. ClearElementFunc happens after (4) in your list.

Reply:
MockElementFactory.ClearElementFunc gets called in `RecycleElementCore` which still happens before we are clearing the DataContext. `RecycleElementCore` gets called by the [this line](https://github.com/chingucoding/microsoft-ui-xaml/blob/0664f291ad0cc0bd4e4bc67aea8f9302c0b04a59/dev/Repeater/ViewManager.cpp#L125) while we are clearing the DataContext in [in this line](https://github.com/chingucoding/microsoft-ui-xaml/blob/0664f291ad0cc0bd4e4bc67aea8f9302c0b04a59/dev/Repeater/ViewManager.cpp#L152).

Reply:
I see. I think we can set DataContext before it goes back to the factory by moving 
```
auto virtInfo = ItemsRepeater::GetVirtualizationInfo(element);
    virtInfo->MoveOwnershipToElementFactory();

    // During creation of this object, we were the one setting the DataContext, so clear it now.
    if (virtInfo->MustClearDataContext())
    {
        if (const auto elementAsFE = element.try_as<winrt::FrameworkElement>())
        {
            elementAsFE.DataContext(nullptr);
        }
    }
```
to [this line.](https://github.com/chingucoding/microsoft-ui-xaml/blob/0664f291ad0cc0bd4e4bc67aea8f9302c0b04a59/dev/Repeater/ViewManager.cpp#L112)

Reply:
I've updated the order now and cleaned up the test as you suggested.